src/Dialect/ONNX/Rewrite.cpp

```diff
@@ -132,6 +132,395 @@ DenseElementsAttr createDenseElementsAttrFromShapeOp(
+    transposedType = RankedTensorType::get(transposedShape, elementType);
+  } else {
+    transposedType = UnrankedTensorType::get(elementType);
+  }
+
+  ONNXTransposeOp transposedInput = rewriter.create<ONNXTransposeOp>(
+      loc, transposedType, x, rewriter.getI64ArrayAttr(perms));
+  return transposedInput.getResult();
+}
+
+// Create ONNX ReverseSequence op
+Value emitONNXReverseSequence(Location loc, PatternRewriter &rewriter,
+    Value input, ArrayRef<int64_t> slens, int64_t batchAxis, int64_t timeAxis) {
+  // Create sequence_lens using Constant op
+  SmallVector<int64_t, 4> dims(1, slens.size());
+  Type tensorType = RankedTensorType::get(dims, rewriter.getIntegerType(64));
+  DenseElementsAttr denseAttr = DenseElementsAttr::get(tensorType, slens);
+  Value constSlens =
+      rewriter.create<ONNXConstantOp>(loc, Attribute(), denseAttr);
+  // Create batch_axis and time_axis attributes
+  IntegerAttr batchAxisAttr =
+      IntegerAttr::get(rewriter.getIntegerType(64, /*isSigned=*/true),
+          APInt(64, batchAxis, /*isSigned=*/true));
+  IntegerAttr timeAxisAttr =
+      IntegerAttr::get(rewriter.getIntegerType(64, /*isSigned=*/true),
+          APInt(64, timeAxis, /*isSigned=*/true));
+  Type resultType = input.getType().cast<RankedTensorType>();
+  ONNXReverseSequenceOp reverseOp = rewriter.create<ONNXReverseSequenceOp>(
+      loc, resultType, input, constSlens, batchAxisAttr, timeAxisAttr);
+  return reverseOp;
+}
+
+// Reverse all elements of the first or second dimention of `input`.
+Value reverseAllElements(
+    PatternRewriter &rewriter, Location loc, Value input, int64_t dimension) {
+  ShapedType inputType = input.getType().cast<ShapedType>();
+  ArrayRef<int64_t> inputShape = inputType.getShape();
+  SmallVector<int64_t, 4> slens;
+  assert((dimension == 0 or dimension == 1) &&
+         "Reversed diemsnion need to be 0 or 1.");
+  // Create `sequence_lengths`, `batch_axis` and `time_axis` to reverse all
+  // elemtns. When reverse the first dim of input(d0 x d1), set `batch_axis` =
+  // 1, and `time_axis` = 0 and create [d0, d0,...,d0] as `sequence_lengths`
+  // whose the number of elements are d1.
+  // Example:
+  //   input(d0 x d1) = (4 x 3)) then, `sequence_lenghts` is [4, 4, 4].
+  // When reverse the second dim of input(d0 x d1), set `batch_axis` = 0,
+  // and `time_axis` = 1 and create [d1, d1,...,d1] as `sequence_lengths`
+  // whose the number of elements are d0.
+  // Example:
+  // input(d0 x d1) = (4 x 3)) then, `sequence_lenghts` is [3, 3, 3, 3].
+  int64_t batch_axis = dimension == 0 ? 1 : 0;
+  int64_t time_axis = dimension == 0 ? 0 : 1;
+  for (int i = 0; i < inputShape[batch_axis]; ++i)
+    slens.emplace_back(inputShape[time_axis]);
+  return emitONNXReverseSequence(
+      loc, rewriter, input, slens, batch_axis, time_axis);
+}
+
+// Reverse elements in weight tensor of ConvTranspose op.
+// The results are used weight tensor for Conv op.
```

Comment:
It would be helpful if you could put an explanation about how is the algorithm for this reverse. I see you scatter them inside the code like ` 1. Transpose ..... 2. Reverse .... 3. Reverse`. Please put them together here so that it is easy to see the whole picture. Thanks!

Reply:
Done. Thanks!